unittests/test_utility.py

```diff
@@ -320,3 +321,287 @@ def test_exception_propagation(self):
+        self.assertEqual(4, scoped_dict['a:b:k3'])
+        self.assertEqual(10, scoped_dict['a:b:k4'])
+
+        self.assertEqual(3, scoped_dict['a:b:c:k1'])
+        self.assertEqual(5, scoped_dict['a:b:c:k2'])
+        self.assertEqual(6, scoped_dict['a:b:c:k3'])
+        self.assertEqual(10, scoped_dict['a:b:c:k4'])
+
+        # Test global scope
+        self.assertEqual(7, scoped_dict['k1'])
+        self.assertRaises(
+            KeyError, exec, "scoped_dict['k2']", globals(), locals()
+        )
+        self.assertEqual(9, scoped_dict['k3'])
+        self.assertEqual(10, scoped_dict['k4'])
+
+        self.assertEqual(7, scoped_dict[':k1'])
+        self.assertRaises(
+            KeyError, exec, "scoped_dict[':k2']", globals(), locals()
+        )
+        self.assertEqual(9, scoped_dict[':k3'])
+        self.assertEqual(10, scoped_dict[':k4'])
+
+        self.assertEqual(7, scoped_dict['*:k1'])
+        self.assertRaises(
+            KeyError, exec, "scoped_dict['*:k2']", globals(), locals()
+        )
+        self.assertEqual(9, scoped_dict['*:k3'])
+        self.assertEqual(10, scoped_dict['*:k4'])
+
+        # Try to fool it, by requesting keys with scope names
+        self.assertRaises(
+            KeyError, exec, "scoped_dict['a']", globals(), locals()
+        )
+        self.assertRaises(
+            KeyError, exec, "scoped_dict['a:b']", globals(), locals()
+        )
+        self.assertRaises(
+            KeyError, exec, "scoped_dict['a:b:c']", globals(), locals()
+        )
+        self.assertRaises(
+            KeyError, exec, "scoped_dict['a:b:c:d']", globals(), locals()
+        )
+        self.assertRaises(
+            KeyError, exec, "scoped_dict['*']", globals(), locals()
+        )
+        self.assertRaises(
+            KeyError, exec, "scoped_dict['']", globals(), locals()
+        )
+
+    def test_setitem(self):
+        scoped_dict = reframe.utility.ScopedDict({
+            'a': {'k1': 1, 'k2': 2},
+            'a:b': {'k1': 3, 'k3': 4},
+            'a:b:c': {'k2': 5, 'k3': 6},
+            '*': {'k1': 7, 'k3': 9, 'k4': 10}
+        })
+
+        scoped_dict['a:k2'] = 20
+        scoped_dict['c:k2'] = 30
+        scoped_dict[':k4'] = 40
+        scoped_dict['*:k5'] = 50
+        scoped_dict['k6'] = 60
+        self.assertEqual(20, scoped_dict['a:k2'])
+        self.assertEqual(30, scoped_dict['c:k2'])
+        self.assertEqual(40, scoped_dict[':k4'])
+        self.assertEqual(50, scoped_dict['k5'])
+        self.assertEqual(60, scoped_dict['k6'])
+
+    def test_delitem(self):
+        scoped_dict = reframe.utility.ScopedDict({
+            'a': {'k1': 1, 'k2': 2},
+            'a:b': {'k1': 3, 'k3': 4},
+            'a:b:c': {'k2': 5, 'k3': 6},
+            '*': {'k1': 7, 'k3': 9, 'k4': 10}
+        })
+
+        # delete key
+        del scoped_dict['a:k1']
+        self.assertEqual(7, scoped_dict['a:k1'])
+
+        # delete a whole scope
+        del scoped_dict['*']
```

Comment:
I think this is equivalent (or almost) to the `test_delete_parent_keep_nested` test. Perhaps, we should merge these two.